game/src/level.rs

```diff
@@ -55,3 +57,30 @@ fn game_level_init(mut commands: Commands, preloaded: Res<PreloadedAssets>) {
         },
     ));
 }
+
+/// attaches parallax components to the different LdtkWorldBundle layers
+fn attach_parallax(
+    mut commands: Commands,
+    query: Query<(Entity, &LayerMetadata, &Transform), Without<Parallax>>,
+) {
```

Comment:
I think this should have an added filter (maybe for `LayerMetadata`) or be run on entering a state, seems like you wouldn't want to run this every update tick as it is now.

Reply:
Hm, I thought by having `&LayerMetaData` as part of the query like that it essentially acts as a filter already though; or maybe you meant something else?

I saw the pattern of just running setup scripts with in update (ab)using the Without filter elsewhere, and thought it would be fine here since the number of LayerMetadata entities is going to be less then 100, so this will never be a perf bottleneck, and it  simplifies some things, since you don't have to worry if there are changes in how layers are spawned in the world or about setup order. (Which seems complex, looking into all the systems that are chained by bevy LDtk, but let me know if there's an easy way to do this I'm overlooking.)

Another option is to add a "DidntAddParallax" marker component, and then filter that out if performance is a concern, but this also adds complexity.

Reply:
Yes, something being part of the query's data access already implies that the query will only match entities that have it. I think odecay meant an `Added` filter, not a `With` filter. Just so that the code of this function only executes when the components are freshly-added.

I call this the "blueprints" pattern. In one system, you insert some component / spawn an entity with an appropriate component. In another system, you have a query with an `Added` filter to perform additional setup / processing.

Reply:
yup thats what i meant